### PR TITLE
Move all report parts to separate files to refactor later

### PR DESCRIPTION
**Pull request by Bisaloo**
This started as a simple PR to address some minor points again. However, I found a thread that I kept pulling and before I knew it, I ended up with https://github.com/epiforecasts/covid19-forecast-hub-europe/commit/2794ec52db83110321f54935771ea515bb1bb7cc. This commit moves the various parts of the reports in separate files (stored in the `rmdchunks/` folder).

It might seem like a complex unnecessary change but I believe this is helpful to refactor the code in reports (#469). A first step in this direction and a good sample of what is possible is the fact I could regroup the forecast plot from `ensemble-report` and from `evaluation-by-country` in a single file (`plot-forecasts.Rmd`) thanks to the addition of a single extra variable (`facetting_var`).

I would like to be able to use more knitr templating features (https://bookdown.org/yihui/rmarkdown-cookbook/knit-expand.html) at some point but the numerous double loops make it a bit difficult at the moment. The only example using this feature at the moment is `template-scores-and-truth-time.Rmd`.

Another important change in the logic is that reports (`ensemble-report.Rmd`, `evaluation-report.Rmd` and `evaluation-by-country.Rmd`) are not completely self-contained and can be rendered without the need to rely on `compute-....-reports.r`. I believe this makes it easier to debug. As an additional bonus, it allowed me to refactor the data loading step in a single chunk `load-data.Rmd`. However, I understand that having to run this step for each country dramatically increases computation time (hence why it was outside the `for` loop and why an external script was necessary) but this is solved by caching this step (https://github.com/epiforecasts/covid19-forecast-hub-europe/commit/be3ddf0a2d556f8992d94d8838bc5b78b61bd3a0) using knitr's built-in caching mechanism.

Another point we can discuss: I believe it makes more sense to call children chunks relative to main file (`../rmdchunks/`) rather than relative to the project root (via `here()`). The rationale is that we might want to change the folder architecture or even their names (e.g. `R-code/` instead of `code/`) and this would break paths specified with `here()` but not paths relative the the main file.

Bonus fix: the DT object no longer overflows on the explanatory text below (see for example Germany evaluation report):

image

### CODE REVIEW

**Review by sbfnk — APPROVED** [review]
I think this is really great and goes a long way towards a more sensible set up. 

The only thing I'm wondering is the extent to which this set up could be used elsewhere, e.g. in another repo that contains forecasts in the same format (e.g., evaluation reports for the forecasts in various folders in https://github.com/epiforecasts/covid-us-forecasts (mainly `models` and `ensembles` I think) or the ensembles in https://github.com/epiforecasts/covid19-forecast-hub-europe/tree/main/ensembles/data-processed). Presumably the main thing that would have to changed is `load-data.Rmd` and/or `define-targets-horizons.Rmd`, and then one could use the same building blocks? Can the child-knitting in Rmarkdown deal with URLs if this is called from another repo? Or could this eventually live in a separate R package from which chunks can be loaded?

Maybe the thing to do is to add a brief set of instructions for this case.

Flagging for completeness that this fixes https://github.com/epiforecasts/covid19-forecast-hub-europe-website/issues/9.

**Comment by sbfnk** [comment]
PS: as for relative file paths vs `here` I don't feel strongly and what you say makes sense to me.

File: code/reports/ensemble/ensemble-report.Rmd [hunk 1]
```diff
@@ -30,173 +31,28 @@ include_calibration <- TRUE
...
-setDT(truth)
-truth[, model := NULL]
-setnames(truth, old = c("value"), 
-         new = c("true_value"))
-
-data <- scoringutils::merge_pred_and_obs(forecasts, truth, 
-                                         join = "full")
-
-target_variables <- c(Cases = "inc case", Deaths = "inc death")
+<!--- Chunk to load forecast and truth data and consolidate it in a single df called data. Usually called as the first chunk. This step is cached so make sure to manually invalidate if necessary when testing. -->
```
[the remark this answers is not in the document]
Comment (Bisaloo) [reply code/reports/ensemble/ensemble-report.Rmd:43]: No, you're right, "manually invalidate" doesn't really make sense. I should write "empty the cache" / "delete the cached file" instead.

For context, there is automatic cache invalidation when:

- the content of `load-data.Rmd` changes (even a comment)
- `params$report_date` changes

More info about knitr's cache invalidation mechanism: https://yihui.org/en/2018/06/cache-invalidation/

**Comment by Bisaloo** [comment]
> The only thing I'm wondering is the extent to which this set up could be used elsewhere, e.g. in another repo that contains forecasts in the same format (e.g., evaluation reports for the forecasts in various folders in https://github.com/epiforecasts/covid-us-forecasts (mainly models and ensembles I think) or the ensembles in https://github.com/epiforecasts/covid19-forecast-hub-europe/tree/main/ensembles/data-processed). Presumably the main thing that would have to changed is load-data.Rmd and/or define-targets-horizons.Rmd, and then one could use the same building blocks? Can the child-knitting in Rmarkdown deal with URLs if this is called from another repo? Or could this eventually live in a separate R package from which chunks can be loaded?

This is something I always keep in mind but at the moment, I'd say that we'll have to wait a little bit until we are able to figure this out. I think there are other parts of the hub structure we have to figure out until we can make any decision about the best way to make these reports re-usable.
At the moment, I think we should focus on adding more things to the config file, upstreaming changes (to `covidHubUtils`, `zoltr`, `scoringutils`, .etc), solidifying choices regarding file formats, etc.
Once we have fixed all these, it should hopefully become clearer how we can generalize the rest / make it more re-usable.

**Comment by nikosbosse** [comment]
This is really nice! 

Two other places where there are essentially duplications of this codes is in github.com/epiforecasts/europe-covid-forecast in the evaluation reports for the crowd forecasts (epiforecasts.io/crowd-evaluation) and epiforecasts.io/uk-challenge. Maybe for these it would be quite easy to go with a similar structure you set up right now and maybe put the templates in a package at some point.

**Comment by sbfnk** [comment]
A simpler interim solution might be to put the report templates in a submodule.

**Comment by sbfnk** [comment]
Alternatively (and perhaps more simply), they could live in the `inst` directory of an R package if going down the route of https://github.com/epiforecasts/covid19-forecast-hub-europe/issues/607.

**Review by kathsherratt — APPROVED** [review]

**Comment by kathsherratt** [comment]
This is great and very needed! 💯 In general I am also keen on packaging this up but I guess that will be a separate project, so definitely makes sense to merge now. (Or maybe to be cautious just after today's report has compiled in case of any issues, since it's used later today.)

**Pull request merged**